### ISSUE TITLE
Issues verifying certificate connecting with HTTPS

### ISSUE THREAD

**Issue by theodorton** [crystal-lang/crystal#5266]
### Crystal version

```
~/code/nabobil/autotoll · (master±)
⟩ crystal -v
Crystal 0.23.1 (2017-10-12) LLVM 4.0.1
```

### How to reproduce

```crystal
# test.cr
require "http/client"

client = HTTP::Client.new("maps.googleapis.com", tls: true)
client.get "/"
```

Then run `crystal run test.cr`.

### Expected outcome

No errors raised, or some redirect.

### Actual outcome

```
SSL_connect: error:14090086:SSL routines:ssl3_get_server_certificate:certificate verify failed (OpenSSL::SSL::Error)
0x10d990e45: *CallStack::unwind:Array(Pointer(Void)) at ??
0x10d990de1: *CallStack#initialize:Array(Pointer(Void)) at ??
0x10d990db8: *CallStack::new:CallStack at ??
0x10d98d615: *raise<OpenSSL::SSL::Error>:NoReturn at ??
0x10da33571: *OpenSSL::SSL::Socket::Client#initialize<TCPSocket, OpenSSL::SSL::Context::Client, Bool, String>:Nil at ??
0x10da333ec: *OpenSSL::SSL::Socket::Client::new:context:sync_close:hostname<TCPSocket, OpenSSL::SSL::Context::Client, Bool, String>:OpenSSL::SSL::Socket::Client at ??
0x10da29c91: *HTTP::Client#socket:(OpenSSL::SSL::Socket+ | TCPSocket+) at ??
0x10da299eb: *HTTP::Client#exec_internal_single<HTTP::Request>:(HTTP::Client::Response | Nil) at ??
0x10da26f7c: *HTTP::Client#exec_internal<HTTP::Request>:HTTP::Client::Response at ??
0x10da26e72: *HTTP::Client#exec<HTTP::Request>:HTTP::Client::Response at ??
0x10da26bdd: *HTTP::Client#exec<String, String, Nil, Nil>:HTTP::Client::Response at ??
0x10da26bb7: *HTTP::Client#get<String>:HTTP::Client::Response at ??
0x10d97bcba: __crystal_main at ??
0x10d98c678: main at ??
```

### Notes

- This also happens when running the code within the official docker image for `0.23.1`.
- I've managed to get it to pass randomly, but I believe there is some cipher requirement and that the certificates differ somewhat between the servers behind this domain
- When running `openssl s_client -connect maps.googleapis.com:443 -tls1_2` I get:

```
CONNECTED(00000003)
depth=2 C = US, O = GeoTrust Inc., CN = GeoTrust Global CA
verify return:1
depth=1 C = US, O = Google Inc, CN = Google Internet Authority G2
verify return:1
depth=0 C = US, ST = California, L = Mountain View, O = Google Inc, CN = *.googleapis.com
verify return:1
---
...
...
```

- It works if I setup a similar example in Ruby:

```ruby
require 'net/http'
require 'uri'

uri = URI('https://maps.googleapis.com/')

Net::HTTP.start(uri.host, uri.port, use_ssl: true) do |http|
  request = Net::HTTP::Get.new uri
  response = http.request request # Net::HTTPResponse object
  puts response
end
```

- In other words, I don't believe this is an issue with my `openssl` configuration.

**Comment by Sija** [comment]
Duplicate of #3477.

See also https://github.com/composer/composer/issues/3346 and [relevant question on SO](https://stackoverflow.com/questions/3298054/ssl-connect-produces-certificate-verify-failure).

**Comment by theodorton** [comment]
I'm not sure, I'm able to connect with TLS to `github.com`:

```
HTTP::Client.get("https://github.com/") # This works
HTTP::Client.get("https://maps.googleapis.com/") # This doesn't
```

I belive there might be something with the allowed cipher suites as [SSLLabs Test results for maps.googleapis.com](https://www.ssllabs.com/ssltest/analyze.html?d=maps.googleapis.com&s=216.58.194.202) show that the root certificate is insecure.

**Comment by asterite** [comment]
I face this problem with some URLs. I think it depends on the version of openssl. Ruby ships with one version of openssl and that's very good: reproducible builds and behavior no matter what your system is. In Crystal we dynamically link to the openssl that's present in the system and it leads to these random and hard to reproduce problems (because we have to have your version of openssl in our machines).

I wish we could just ship Crystal with one version of openssl, one version of llvm, one version of libpcre, one version of libxml2, etc., so that everything works exactly the same everywhere...

**Comment by RX14** [comment]
@asterite And then what if openssl has a security bug? If we distribute our own libraries we are now distro maintainers. We have to follow the security bugs closely with a quick response, and possibly backport patches, and deal with a whole bunch of things we don'y have time to do. This is the distro maintainer's job, not ours, it's already done for us.

I've never had an openssl problem with crystal on archlinux, i've only seem people have problems on osx and on debian-based (read: ancient versions of openssl) distros.

**Comment by theodorton** [comment]
I don't have any strong opinions on whether to bundle openssl with the compiler or not, but I think this doing `VERIFY_NONE` as a workaround (which is mentioned in the linked issues) here unsatisfactory.

I tried to use `OpenSSL::SSL::Context::Client#add_x509_verify_flags(OpenSSL::SSL::X509VerifyFlags:: TRUSTED_FIRST)` which was mentioned as a fix in a Stack Overflow issue I found yesterday. I'm unable to find it again now, but the problem I ran into was that `OpenSSL::SSL::X509VerifyFlags` was not defined. Seems like there's a compiler flag for that constant that wasn't set.

It seems highly related, but I'm not sure? I'm building Crystal via Homebrew. Could it be that it's missing some SSL/crypto lib while compiling? Or that the compiled bottle comes without this support?

**Comment by RX14** [comment]
It appears to me that openssl's API is a mess, perhaps we should implement the `libtls` API in C using openssl, how openssl wants us to, and then use that as the base for crystal's SSL.

**Comment by theodorton** [comment]
The workaround here is:

```
context = OpenSSL::SSL::Context::Client.insecure
client = HTTP::Client.new("maps.googleapis.com", tls: context)
```

I'm not sure if that's a very good workaround though, as if I point the client to `expired.badssl.com`, I'm not getting any errors.

**Comment by ysbaddaden** [comment]
Hence the `insecure`: you basically accept whatever OpenSSL supports —I'm not even sure it verifies certificates. Would you be willing to take the `Client.insecure` and tweak it until you find the culprit in defaults?

**Comment by theodorton** [comment]
@ysbaddaden Yes, we're ok with `Client.insecure` for now, but it doesn't seem to verify certificates. I think that should probably be noted in the documentation. Currently the docs say:

> Use this only if undoing the defaults that new sets is too much hassle.

Not sure if I can spend more time at the moment looking into the defaults. I'm not very fluent in TLS terminology so I'd just have to guess what the different options represent. The failures were pretty consistent with that hostname, even when I ran it in the Docker image for `0.23.1`.

I still think this is highly relevant:

> I'm still getting undefined constant OpenSSL::OPENSSL_102 errors when attempting to use X509VerifyFlags.

As I believe using the `OpenSSL::SSL::X509VerifyFlags:: TRUSTED_FIRST` might work.

**Comment by vladfaust** [comment]
Any updates? I'm basically unable to connect to any of Google's APIs :confused: 

```crystal
require "http/client"
client = HTTP::Client.new(URI.parse("https://google.com"))
p client.get("/")
```

It raises the same `certificate verify failed (OpenSSL::SSL::Error)`.

**Comment by will** [comment]
Just chiming into say that I'm having this problem on heroku with google oauth.

Also since the oauth::client doesn't take a tls: context, the workaround of not verifying doesn't work, unless you monkey patch

**Comment by waj** [comment]
I just sent a PR with a potential fix. Please, let me know if it works for you and also provide some comments on the solution if possible.

**Comment by codenoid** [comment]
@waj @ysbaddaden @RX14

**Comment by RX14** [comment]
@codenoid it's because you specify `Accept-Encoding` in the request headers. You're telling crystal you accept the gzip encoding, and so crystal gives you back your gzip encoding.

Please don't copy your browser's headers into crystal code, just 
```cr
require "http"

HTTP::Client.get("https://maps.googleapis.com/maps/api/geocode/json?&address=Kalibata%20City")
```

is enough and you get a nice response by default.

**Comment by codenoid** [comment]
wat, why it's work ? because when i try with simple GET, i got 400 :/ aarrrg

**Comment by sam0x17** [comment]
I am also experiencing this issue here in the context of a docker image running in lambda via up.sh: https://github.com/Sija/raven.cr/issues/47#issuecomment-520280743

**Comment by benbonnet** [comment]
Hello all;

quite new to crystal; focusing on getting the smallest docker image (so, statically building a crystal app).

Currently stuck on the issue here; I was thinking about using an intermediate to retrieve the required certs :

```
FROM crystallang/crystal:latest as compiler
RUN mkdir /app
COPY . /app
RUN mkdir /app/bin
WORKDIR /app
RUN shards --production
RUN crystal build /app/src/appname.cr --static --release --no-debug -o /app/bin/appname

FROM alpine as certs-builder
RUN apk update && apk upgrade && apk add --no-cache ca-certificates
RUN update-ca-certificates

FROM busybox:glibc
COPY --from=compiler /app/bin/appname .
COPY --from=certs-builder /etc/ssl/certs /etc/ssl/certs

CMD ["./appname"]
```

Would there be a way to achieve something similar by specifying the path to the certs in the tls context ? Or should I choose another approach ?

**Comment by Blacksmoke16** [comment]
@benbonnet The problem is statically linking is only supported in Alpine linux. See https://crystal-lang.org/reference/using_the_compiler/index.html#creating-a-statically-linked-executable and https://crystal-lang.org/2020/02/02/alpine-based-docker-images.html.

**Comment by benbonnet** [comment]
@Blacksmoke16
As shown above it is first built within `crystallang/crystal:latest`.  
On the final step; the app bin was copied through a `FROM scratch` in the last step.  
Had to move on and use `FROM busybox:glibc` (otherwise it could not resolve hostnames; among other problems).  
As this slight change enabled a missing capability; am in hope to find a way to just get what's needed be able to verify the certs

**fully** statically linked are only possible within alpine; but could the same can be achieved by picking only the needed requirements and putting those in the final image ?

**Comment by Blacksmoke16** [comment]
@benbonnet I guess what I'm getting at is couldn't you statically link the binary in the Crystal alpine image, then move it to another smaller Alpine image and just use that?  Like I suggested in https://github.com/crystal-lang/crystal/issues/6099#issuecomment-753382863?

**Comment by benbonnet** [comment]
@Blacksmoke16 well you might get a way larger image (10x maybe ?). I do know my asking is completely overkill and we would still be fine by not even building statically.  
But I've been super hyped to see how the smallest image possible would reduce our pipeline deployment and functions startup (we're using cloudrun, such image cold starts in 200ms or less)

**Comment by Blacksmoke16** [comment]
@benbonnet

```sh
docker images
REPOSITORY       TAG                         IMAGE ID            CREATED             SIZE
busybox          glibc                       b0748d3617e4        9 days ago          5.21MB
alpine           3.12                        389fef711851        3 weeks ago         5.57MB
```

Size wise they're essentially the same, but with alpine you have a package manager and such.  Granted this is not including the size of your Crystal binary, or other dependencies you may need.

**Comment by straight-shoota** [comment]
> Still failing.

What exactly is failing? Is it related to this issue or a different one? From the static linking discussion it feels like something else, which would probably fit better to https://forum.crystal-lang.org/

Not sure if dynamic linking is a factor for your issue, but when you copy a dynamically linked binary to a different system, you need to make sure it has matching dynamic libraries available.
I would strongly recommend to link fully statically (can use the alpine image for that). That causes less friction with moving between systems and should help you goal to reach minimal binary size.
